packages/opentelemetry-exporter-datadog/src/datadogSpanProcessor.ts

```
@@ -0,0 +1,203 @@
+      return;
+    }
+
+    const traceId = span.spanContext['traceId'];
+
+    if (!this._traces.has(traceId)) {
+      this._traces.set(traceId, []);
+      // this._traces.get(traceId).set(SPANS_KEY, [])
+      this._traces_spans_started.set(traceId, 0);
+      this._traces_spans_finished.set(traceId, 0);
+    }
+
+    const trace = this._traces.get(traceId);
+    const traceSpansStarted = this._traces_spans_started.get(traceId);
+
+    if (traceSpansStarted >= this._maxTraceSize) {
+      this.logger.debug('Max Trace Size reached, dropping span');
+      return;
+    }
+
+    trace.push(span);
+    this._traces_spans_started.set(traceId, traceSpansStarted + 1);
+  }
+
+  onEnd(span: ReadableSpan): void {
+    if (this._isShutdown) {
+      return;
+    }
+
+    const traceId = span.spanContext['traceId'];
+
+    if (!this._traces.has(traceId)) {
+      return;
+    }
+
+    const traceSpansFinished = this._traces_spans_finished.get(traceId);
+
+    this._traces_spans_finished.set(traceId, traceSpansFinished + 1);
+
+    if (this._isExportable(traceId)) {
+      this._check_traces_queue.add(traceId);
+      this._maybeStartTimer();
+    }
+  }
+
+  getTraceContext(span: ReadableSpan): string | undefined[] {
+    return [
+      id(span.spanContext['traceId']),
+      id(span.spanContext['spanId']),
+      span.parentSpanId ? id(span.parentSpanId) : undefined,
+    ];
+  }
+
+  /** Send the span data list to exporter */
+  private _flush() {
+    this._clearTimer();
+    if (this._check_traces_queue.size === 0) return;
+
+    this._check_traces_queue.forEach(traceId => {
+      // check again in case spans have been added
+      if (this._isExportable(traceId)) {
+        const spans = this._traces.get(traceId);
+        this._traces.delete(traceId);
+        this._traces_spans_started.delete(traceId);
+        this._traces_spans_finished.delete(traceId);
+        this._check_traces_queue.delete(traceId);
+        this._exporter.export(spans, () => {});
```

you might want to log an error here

---

I agree this will make an export fail silently which is no good

---

yup makes sense, i've added an error log here. (are there preferences on log level? error feels appropriate here since most debug logs wont get registered, but not sure if there's otel conventions are noiseyness)

---

The exporter for the otel collector logs an error for reference, so i think thats the way to go

---

ty, will keep as error log.